[Bat] Mission3- 체스판 초기화

kjk402 opened the pull request:
안녕하세요, 웹백엔드과정 Bat입니다.

하기와 같이 미션3 구현하여 pull request 드립니다.

• mission3 summary
=====================================
• commit message template 사용.
이번 미션에서 commit을 할 때 자세한 message를 남기고자 template을 사용하였습니다.
Add initialize method in Board.java  를 commit하고 한글이 깨진 것을 확인하여 . gitconfig에서 Encoding부분을 utf-8로 변경 후 정상적으로 한글 업로드 확인하였습니다.

• Main에서 (else 문 사용을 하지 않기위해) continue와 break를 사용하여 특정 명령어가 아닌 다른 명령어를 입력받았을 때의 동작을 구현하였습니다. 

• Pawn의 color가 white, black 두 가지가 있습니다. 또, color에 따라 representation을 white면 'p'
 black일 시 'P'을 나타내며 변경되지 않는 값으로 final 선언해주었습니다.

=====================================

• 지난 미션에서의 개선점
접근제한자를 신경써서 코딩하였습니다.
DisplayName을 자세히 작성하였습니다.

• mission3을 완료 후 인지한 점.
 Board 클래스의 print()에서 줄바꿈 \n 을 사용하지 않고 다른 방법인 appendNewLine을 사용하기. (mission4 요구사항이라 아직 적용하지 않았습니다.)


리뷰 잘 부탁드립니다.
감사합니다.

kjk402 committed:
Pawn.java

Pawn.java �¦ arrayList White, Black �庖¦

kjk402 committed:
Add initialize method in Board.java
ì²스판에서 pawn 8개 추가해주는 메소드
생성

kjk402 committed:
Add initialize in BoardTest

BoardTest에서 initialize는 흰색폰과 검은색폰의 생성된 갯수가 알맞은 지
확인한다.
Board.java에서는 list에 담긴 객체를 SrtingBuilder에 담아 리턴합니다.

kjk402 committed:
print메소드 구현

StringBuilder boardOutput을 선언하여 체스판을 담고, String으로 변환하는
toString()을 사용하여 출력한다.

kjk402 committed:
Create Main Class

main() 메소드를 생성하여 Board에서의 initialize와  print가
정상적으로 동작하는 지 확인.

kjk402 committed:
main()메소드에 시작과 종료기능 추가

while문 안에서 Scanner를 사용하여 입력 값을 받고 start일 때 콘솔에 출력, end면 break로 종료.

kjk402 committed:
Main 클래스에 도움말 추가.

help메소드는 명령어를 리턴해줍니다. 명령어는 변하지 않기에 final로
선언해주었습니다.

kjk402 committed:
PawnTest class 수정

Pawn 클래스의 color와 representation을 받아서 테스트하도록 수정.
DisplayName 추가.

kjk402 committed:
Board 클래스 수정

Main 클래스에서 실행 시 PPPPPPPP 이후 개행되지않고 이어서 출력되어서 \n 추가

kjk402 committed:
Board class 수정

getWhitePawnsResult 와 getBlackPawnsResult의 접근제한자가 생략되어있어 private로
변경 후  정상 동작 확인하였습니다.

kjk402 committed:
Board class 재수정

Board class의 getWhitePawnsResult와 getBlackPawnsResult의 접근제한자를
private로 선언해주었더니,
BoardTest 클래스에서Test 에러 발생으로 public 으로 변경.

SANGCO commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -6,18 +6,67 @@
-    public int size() {
-        return pawnList.size();
+    public void addBlackPawn(Pawn pawn) {
+        blackPawnList.add(pawn);
```
addWhitePawn() 메서드에 검정색 Pawn을 넣는 테스트를 하나 만들어봅시다!
그리고 그런 상황에서는 어떻게 해야할지 코드를 수정해 봅시다~

kjk402 replied:
해당 피드백을 참고하여 @Test addBlackPawnInWhitePawnList() 을 생성하였을 때는 정상 작동했습니다. 
그래서 addWhitePawn과 addBlackPawn에서 color에 따라 add 되도록 수정하였습니다.
수정 이후에는 @Test 오류나는 것 확인하였습니다.

SANGCO commented on src/main/java/net/kjk402/chess/chess/Main.java:
```diff
@@ -0,0 +1,36 @@
+
+public class Main {
+    public static void main(String[] args) {
+        Scanner sc = new Scanner(System.in);
```
Scanner와 같이 AutoCloseable을 구현한 클래스의 경우 
[try-with-resources](https://www.baeldung.com/java-try-with-resources)를 사용해서 쉽게 리소스를 해제할 수 있답니다.
이번 미션에서 try-with-resources 관련되어서 학습하고 적용해보면 좋을거 같아요~

kjk402 replied:
해당 피드백으로  try-with-resources 적용하여 수정하였습니다.

감사합니다. ^^

SANGCO commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -6,18 +6,67 @@
+        return sb.toString();
+    }
+
+    void print() {
```
틀별히 디퐅트 접근 제한자를 선택하신 이유는?

kjk402 replied:
다른 패키지에서는 print 사용하지 않고 같은 패키지인 Main에서만 사용할 것 같아 디폴트로 하였습니다.

SANGCO replied:
접근 제한자를 선택하실 때  public 혹은 private 만들고
private인데 상속 관계에 있는 객체들에서 사용해야 한다고 하면 protected로
private 인데 다른 패키지에서 사용을 해야 한다면 package-private(default)로 
이런식으로 접근을 하면 될거 같은데
호눅스 혹시 다른 의견이 있으시면 댓글 부탁드립니다ㅋ

kjk402 replied:
상속받은 클래스에서만 접근 가능하도록 protected 접근 제한자로 수정하였습니다.

SANGCO replied:
public으로 했다가 실제로 필요할 때 protected로 바꾸는건 어떨까요~

SANGCO commented on src/main/java/net/kjk402/chess/chess/Main.java:
```diff
@@ -0,0 +1,36 @@
+        sc.close();
+    }
+
+    private static final String Help() {
```
메소드 이름이 Help() 이렇게 대문자로 되어있군요!
help() 메서드에 final을 붙이신 이유는?
마지막으로 메서드 명에 대해서 한번더 고민해 보면서 
이 메서드가 꼭 필요한가에 대해서도 생각해 봅시다ㅋ

kjk402 replied:
메소드명 대문자에서 소문자로 수정하였습니다.

final 을 선언한 이유는  명령어는 불변, 접근이 불가능해야 된다고 생각하여 다른 곳에서 재정의할 수 없도록 final을 붙였습니다. 
final을 이런 용도로 사용하는 것이 틀린 것인지 궁금합니다.

kjk402 replied:
help메소드는 도움말 개념으로 사용되며, 이후 명령어가 추가될 가능성을 생각하여 따로 메소드로 만들었습니다.

SANGCO replied:
final이 클래스, 필드, 메서드 어디에 붙는지에 따라 기능이 다른데요.
이 부분을 이번에 확실하게 정리하고 넘어가시죠!

그리고 현재 help() 메서드는 너무 역할이 없는거 같은데
향후에 뭔가 역할이 생기면 그때 메서드로 빼도 되지 않을까하고 생각합니다~

kjk402 replied:
네. 제 생각에도 help 메서드의 역할이 적은 것 같습니다.
메서드로 따로 빼지 않고 main으로 수정하였습니다.

감사합니다.~~

SANGCO commented on src/main/java/net/kjk402/chess/pieces/Pawn.java:
```diff
@@ -1,20 +1,29 @@
-    private String color;
+    private final String color;
+    private final char representation;
     public static final String WHITE = "white";
     public static final String BLACK = "black";
+    public static final char WHITE_REPRESENTATION = 'p';
+    public static final char BLACK_REPRESENTATION = 'P';
```
static이 붙은 클래스 변수를 인스턴스 변수 위로 위치를 변경해 주세요~

kjk402 replied:
네. 해당 부분 수정하였습니다.

클래스 변수를 인스턴스 변수 위쪽으로 옮긴 이유가
static 붙은 클래스 변수는 클래스 로드 시에 가장 먼저 초기화 되기에 위쪽에 위치한다. 
라는 이유인데 혹시 제가 생각한 것이 맞을까요?

SANGCO replied:
[3.1.3 Class and Interface Declarations](https://www.oracle.com/java/technologies/javase/codeconventions-fileorganization.html)

뭘 먼저 선언 해야는지 순서가 또 있거든요.
링크 참고해 주세요~

kjk402 replied:
링크 감사합니다. ~

SANGCO reviewed (changes requested):
bat님 구현 잘하셨네요. 고생하셨습니다!
리뷰 남겼으니 확인해 보시고 질문 있으시면 댓글 부탁드립니다~

kjk402 committed:
Main 클래스 수정

Main class에서 Help메소드명 대문자로 되어있는 것 소문자로 변경

kjk402 committed:
Pawn class 수정

클래스변수와 인스턴스변수를 합쳐 멤버변수라 불리는데 멤버변수 생성
시에 클래스변수가 인스턴스 변수 위에 위치해야한다.

kjk402 committed:
Use try-with-resources to fix Main class

Scanner 사용 시 자동으로 close 기능을 제공하는 try-with-resources
를 사용하여 수정하였습니다.

kjk402 committed:
Board class, BoardTest class 수정(test추가)

이전에는 검정색 폰을 addWhitePawn 메서드를 사용해도 테스트 이상이
없었는데, if문을 사용하여 color가 white일 때만 add 되도록
수정하였습니다.

kjk402 committed:
Main class 수정

help (도움말)가 복잡하지 않아  굳이 메서드로 뺄 필요없다는 피드백을
주셔서 수정하였습니다.

kjk402 committed:
Use protected for print() in Board class

protected 접근 제한자를 붙이면 같은 패키지의 상속관계에 있는 클래스 에서만 사용 가능합니다.

SANGCO commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -6,18 +6,71 @@
+        return sb.toString();
+    }
+
+    protected void print() {
```
위에 댓글 남겼는데 public으로 했다가 실제로 필요할 때 protected로 바꾸는게 좋을거 같네요.

kjk402 replied:
public으로 다시 고쳤습니다.

SANGCO commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -6,18 +6,71 @@
-        pawnList.add(pawn);
+    public void addWhitePawn(Pawn pawn) {
+        if (pawn.getColor() == "white") {
+            whitePawnList.add(pawn);
+        }
     }
 
-    public int size() {
-        return pawnList.size();
+    public void addBlackPawn(Pawn pawn) {
+        if (pawn.getColor() == "black") {
+            blackPawnList.add(pawn);
+        }
```
"white" 스트링 사용하기 보다 만들어 놓은 상수를 사용해 봅시다.

kjk402 replied:
"white" 랑 "black" 을 Pawn에서 선언한 상수인 Pawn.WHITE, Pawn.BLACK 으로 수정하였습니다.

SANGCO commented on src/main/java/net/kjk402/chess/chess/Main.java:
```diff
@@ -0,0 +1,35 @@
+        StringBuilder help = new StringBuilder();
+        help.append("시작명령어: start\n");
+        help.append("종료명령어: end\n");
+        help.toString();
```
`help.toString();` 이거는 필요없는 코드군요.

kjk402 replied:
넵 해당 코드 삭제했습니다.

SANGCO commented on src/main/java/net/kjk402/chess/chess/Main.java:
```diff
@@ -0,0 +1,35 @@
+                if (command.equals("end")) {
+                    System.out.println("종료합니다.");
+                }
+                if (command.equals("help")) {
+                    System.out.println(help);
+                }
```
왜 메서드 이름이 help()지 했는데
이런 기능이었군요ㅋ

SANGCO reviewed (changes requested):
간단한 리뷰 추가합니다~

kjk402 committed:
Fix Main, Board class

Main class print() 메서드 public으로 변환, toString 삭제.
Board class addWhitePawn, addBlackPawn의 if문에서 스트링이 아닌
만들어놓은 상수로 수정.

SANGCO reviewed (approved):
고생하셨습니다~

kjk402 committed:
Git encoding file 수정

commit 시에 한글 깨지는 오류 해당 설정으로 해결.

honux77 commented:
👍 두 분 다 수고하셨어요!

kjk402 commented:
SANGCO, honux 리뷰해주셔서 감사합니다.
